llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1207,6 +1207,48 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
     }
   }
 
+  // (X * Y) / (X * Z) --> Y / X (and commuted variants)
+  if (match(Op0, m_Mul(m_Value(X), m_Value(Y)))) {
+    auto OB0HasNSW = cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap();
+    auto OB0HasNUW = cast<OverflowingBinaryOperator>(Op0)->hasNoUnsignedWrap();
+
+    auto CreateDivOrNull = [&](Value *A, Value *B) -> Instruction * {
+      auto OB1HasNSW = cast<OverflowingBinaryOperator>(Op1)->hasNoSignedWrap();
+      auto OB1HasNUW =
+          cast<OverflowingBinaryOperator>(Op1)->hasNoUnsignedWrap();
+      const APInt *C;
+      if (IsSigned && OB0HasNSW) {
+        if (OB1HasNSW && (match(B, m_APInt(C))) && (!C->isAllOnes()))
+          return BinaryOperator::CreateSDiv(A, B);
+        if (auto *SV = simplifySDivInst(A, B, /*IsExact*/ true,
+                                        SQ.getWithInstruction(&I))) {
+          if (auto *CV = dyn_cast<Constant>(SV);
+              CV && !(isa<PoisonValue>(CV) || CV->containsPoisonElement()))
+            if (match(B, m_APInt(C)) && !C->isSignMask())
+              return replaceInstUsesWith(I, SV);
+        }
```

Comment:
```suggestion
        if (match(A, m_APInt(C1)) && match(B, m_APInt(C2)) && !C2->isSignMask()) {
          APInt Q, R;
          APInt::sdivrem(*C1, *C2, Q, R);
          if (R->isZero())
            return replaceInstUsesWith(I, ConstantInt::get(A->getType(), Q));
        }
```
I think this implementation is simpler.